master/internal/db/postgres_groups.go

```diff
@@ -0,0 +1,218 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/jackc/pgconn"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+// AddGroup adds a group to the database. Returns ErrDuplicateRow if a
+// group already exists with the same name or ID.
+func (db *PgDB) AddGroup(ctx context.Context, group model.Group) (model.Group, error) {
+	_, err := Bun().NewInsert().Model(&group).Exec(ctx)
+	return group, matchSentinelError(err)
+}
+
+// GroupByID looks for a group by id. Returns ErrNotFound if the group isn't found.
+func (db *PgDB) GroupByID(ctx context.Context, gid int) (model.Group, error) {
+	var g model.Group
+	err := Bun().NewSelect().Model(&g).Where("id = ?", gid).Scan(ctx)
+
+	return g, matchSentinelError(err)
+}
+
+// SearchGroups searches the database for groups. userBelongsTo is "optional"
+// in that if a value < 1 is passed in, the parameter is ignored. SearchGroups
+// does not return an error if no groups are found, as that is considered a
+// successful search.
+func (db *PgDB) SearchGroups(ctx context.Context,
+	userBelongsTo model.UserID) ([]model.Group, error) {
+	var groups []model.Group
+	query := Bun().NewSelect().Model(&groups)
+
+	if userBelongsTo > 0 {
+		query = query.
+			Join("INNER JOIN user_group_membership AS ugm ON ugm.group_id=groups.id").
+			Where("ugm.user_id = ?", userBelongsTo)
+	}
+
+	err := query.Scan(ctx)
+
+	return groups, err
+}
+
+// DeleteGroup deletes a group from the database. Returns ErrNotFound if the
+// group doesn't exist.
+func (db *PgDB) DeleteGroup(ctx context.Context, gid int) error {
```

Comment:
can we have `ON DELETE CASCADE` for foreign keys in `user_group_membership` in SQL instead? then this won't have to be a transaction.